src/properties/forest.c

```diff
@@ -0,0 +1,216 @@
+ *
+ * \param graph The graph object to analyze.
+ * \param res Pointer to a logical variable, the result will be stored
+ *        here.
+ * \param roots A vector in which the root nodes will be stored. When \p mode
+ *        is \c IGRAPH_ALL or the graph is undirected, any 1 vertex from each
+ *        component can be the root. When \p mode is \c IGRAPH_OUT
+ *        or \c IGRAPH_IN, all the vertices with zero in- or out-degree,
+ *        respectively are considered as root nodes.
+ * \param mode For a directed graph this specifies whether to test for an
+ *        out-tree, an in-tree or ignore edge directions. The respective
+ *        possible values are:
+ *        \c IGRAPH_OUT, \c IGRAPH_IN, \c IGRAPH_ALL. This argument is
+ *        ignored for undirected graphs.
+ * \return Error code:
+ *        \c IGRAPH_EINVMODE: invalid mode argument.
+ *
+ * Time complexity: At most O(|V|+|E|), the
+ * number of vertices plus the number of edges in the graph.
+ *
+ */
+igraph_error_t igraph_is_forest(const igraph_t *graph,igraph_bool_t *res, igraph_vector_t *roots , igraph_neimode_t mode) {
+    igraph_adjlist_t al;
+    igraph_vector_t visited;
+    igraph_integer_t visited_count=0;
+    igraph_integer_t vcount, ecount;
+
+    vcount = igraph_vcount(graph);
+    ecount = igraph_ecount(graph);
+    IGRAPH_CHECK(igraph_vector_resize(roots,0));
+
+    /*By convention, a zero-vertex graph will be considered a forest.*/
+    if (vcount == 0) {
+        *res = 1;
+        return IGRAPH_SUCCESS;
+    }
+    // A forest can have maximum vcount-1 edges.
+    if (ecount > vcount - 1) {
+        *res = 0;
+        return IGRAPH_SUCCESS;
+    }
+    /*A single-vertex graph is a forest, provided it has no edges (checked in the previous if (..)) */
+    if (vcount == 1) {
+        *res = 1;
+        igraph_vector_push_back(roots,0);
+        return IGRAPH_SUCCESS;
+    }
+    /* Ignore mode for undirected graphs. */
+    if (! igraph_is_directed(graph)) {
+        mode = IGRAPH_ALL;
+    }
+    IGRAPH_CHECK(igraph_adjlist_init(graph, &al, mode, IGRAPH_LOOPS, IGRAPH_MULTIPLE));
+    IGRAPH_FINALLY(igraph_adjlist_destroy, &al);
+
+    *res = 1; /* assume success */
+    IGRAPH_CHECK(igraph_vector_init(&visited,vcount));
+    IGRAPH_FINALLY(igraph_vector_destroy, &visited);
+    igraph_integer_t i;
+
+    /* The main algorithm:
+     * Undirected Graph:- We add each unvisited vertex to the roots vector, and
+     * mark all other vertices that are reachable from it as visited.
+     *
+     * Directed Graph:- For each tree, the root is the node with no
+     * incoming/outgoing connections, depending on 'mode'. We add each vertex
+     * with zero degree to the roots vector and mark all other vertices that are
+     * reachable from it as visited.
+     *
+     * If all the vertices are visited exactly once, then the graph is a forest.
+     */
+
+    switch (mode) {
+    case IGRAPH_ALL:{
```

Comment:
In the rest of the codebase, `case` branches are indented by one indentation unit (4 spaces) relative to the `switch` statement. Please update the coding style accordingly.